File: src/main/java/me/mattstudios/triumphpets/pet/nms/v1_14_r1/goals/PathfinderGoalFollowPlayer.java
package me.mattstudios.triumphpets.pet.nms.v1_14_r1.goals;

import me.mattstudios.triumphpets.pet.components.PetMemory;
import net.minecraft.server.v1_14_R1.EntityInsentient;
import net.minecraft.server.v1_14_R1.NavigationAbstract;
import net.minecraft.server.v1_14_R1.PathfinderGoal;
import org.bukkit.Location;
import org.bukkit.craftbukkit.v1_14_R1.entity.CraftPlayer;
import org.bukkit.entity.Player;

import static me.mattstudios.triumphpets.util.Utils.distance2d;
import static me.mattstudios.triumphpets.util.Utils.getSafeY;

public class PathfinderGoalFollowPlayer extends PathfinderGoal {

    private final double MOVEMENT_SPEED;
    private EntityInsentient petEntity;
    private Player owner;
    private NavigationAbstract navigation;
    private PetMemory petMemory;

    private int followDistance;
    private int tpDistance;

    public PathfinderGoalFollowPlayer(EntityInsentient petEntity, Player owner, PetMemory petMemory, double MOVEMENT_SPEED) {
        this.petEntity = petEntity;
        this.owner = owner;
        this.navigation = petEntity.getNavigation();
        this.MOVEMENT_SPEED = MOVEMENT_SPEED;
        this.petMemory = petMemory;

        followDistance = 7;
        tpDistance = 20;
    }

    @Override
    public boolean a() {

        if (petMemory.isTracking()) followDistance = 15;
        else followDistance = 7;

        Location location = owner.getLocation().clone();
        double dist = distance2d(location.getX(), petEntity.locX, location.getZ(), petEntity.locZ);

        if (dist >= followDistance) {
            if (dist >= tpDistance) {
                petEntity.setPosition(location.getX(), getSafeY(location), location.getZ());
            }

            move();
        }

        return false;
    }

    private void move() {
        navigation.a(((CraftPlayer) owner).getHandle(), MOVEMENT_SPEED);
    }

}

File: version1_15/src/main/kotlin/me/mattstudios/triumphpets/pet_1_15_r1/goal/PathfinderGoalFollowPlayer.kt
package me.mattstudios.triumphpets.pet_1_15_r1.goal

import me.mattstudios.triumphpets.pet.Pet
import me.mattstudios.triumphpets.pet.PetUtils.distance2d
import me.mattstudios.triumphpets.pet.PetUtils.getSafeY
import net.minecraft.server.v1_15_R1.EntityInsentient
import net.minecraft.server.v1_15_R1.NavigationAbstract
import net.minecraft.server.v1_15_R1.PathfinderGoal
import org.bukkit.Location
import org.bukkit.craftbukkit.v1_15_R1.entity.CraftPlayer


/**
 * @author Matt
 */
class PathfinderGoalFollowPlayer(pet: Pet, private val entityInsentient: EntityInsentient, private val MOVEMENT_SPEED: Double) : PathfinderGoal() {

    private val petMemory = pet.getMemory()
    private val owner = pet.getOwner()

    private val navigation: NavigationAbstract  = entityInsentient.navigation
    private var followDistance = 7
    private val tpDistance = 20

    override fun a(): Boolean {

        followDistance = if (petMemory.tracking) 15 else 7

        val location: Location = owner.location.clone()
        val dist: Double = distance2d(location.x, entityInsentient.locX(), location.z, entityInsentient.locZ())

        if (dist >= followDistance) {
            if (dist >= tpDistance) {
                entityInsentient.setPosition(location.x, getSafeY(location, owner.world).toDouble(), location.z)
            }

            move()
        }

        return false
    }

    private fun move() {
        navigation.a((owner as CraftPlayer).handle, MOVEMENT_SPEED)
    }
}